include/envoy/stats/symbol_table.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Interface for storing a stat name.
+ */
+class StatName {
+public:
+  virtual ~StatName(){};
+  virtual std::string toString() const PURE;
+};
+
+using StatNamePtr = std::unique_ptr<StatName>;
+
+/**
+ * Interface for shortening and retrieving stat names.
+ *
+ * Guarantees that x = encode(x).toString() for any x.
+ */
+class SymbolTable {
+public:
+  virtual ~SymbolTable() {}
+
+  /**
+   * Encodes a stat name into a StatNamePtr. Expects the name to be period-delimited.
+   *
+   * @param name the stat name to encode.
+   * @return StatNamePtr a unique_ptr to the StatName class encapsulating the symbol vector.
+   */
+  virtual StatNamePtr encode(const absl::string_view name) PURE;
```

Comment:
There is no value in making a `absl::string_view` a `const` on an interface, i.e. this doesn't force any property on the caller or provide them any guarantee. You can (very optionally) use it on the implementation definition if you want.